EE_endpoints.md

```diff
@@ -0,0 +1,130 @@
+
+The sample [Hello
+Phoenix](https://github.com/phoenix-examples/hello_phoenix) Phoenix application starts up our HelloPhoenix.Endpoint as a supervised process.
+This is where Supervision from Erlang and Elixir enter this framework.
+
+Below is from
+[`lib/hello_phoenix.ex`](https://github.com/phoenix-examples/hello_phoenix/blob/master/lib/hello_phoenix.ex#L9-L16):
+```elixir
+ # ...
+ children = [¬
+   # Start the endpoint when the application starts¬
+   supervisor(HelloPhoenix.Endpoint, []),¬
+   # ... Other supervised processes like Ecto Repo or
+   # worker processes
+   supervisor(HelloPhoenix.Repo, []),
+   supervisor(HelloPhoenix.QueueWorker, [])
+  ]
+  # ...
+```
+
+These lines declare that our application is composed of multiple process trees.
+
+### Endpoint Contents
+An endpoint often includes plugs that are useful throughout all routes of
+your application. At the end of an endpoint file we can define multiple routers that would
+handle the actual routes.
+
+Let's cover some of the elements of an Endpoint layer by layer from outside
+moving inward by looking at the endpoint.ex in the [Hello
+Phoenix](https://github.com/phoenix-examples/hello_phoenix) sample Phoenix
+application.
+
+An `Endpoint` is an Elixir module like the `Controller` and `Router`.
+
+```elixir
+defmodule HelloPhoenix.Endpoint do¬
+  ...
+end
+```
+Frequently the first directive inside of our Endpoint module will be an inclusion
+of the Phoenix.Endpoint and a declaration that this is an otp_app. [This
+code](https://github.com/phoenixframework/phoenix/blob/e118c485a1a0bdc1f4f2fe199f980b5fff691376/lib/phoenix/endpoint.ex#L362-L369) is
+executed when the below `use` line is executed. This declares that this is an
+otp_app called "hello_phoenix" which is enough for the OTP configuration to get
+started and [start the endpoint supervision
+tree](https://github.com/phoenixframework/phoenix/blob/e118c485a1a0bdc1f4f2fe199f980b5fff691376/lib/phoenix/endpoint.ex#L467)
+```elixir
+  use Phoenix.Endpoint, otp_app: :hello_phoenix
+```
+Next the endpoint declares a socket on the /socket URI. /socket requests will be handled by the
+HelloPhoenix.UserSocket module which is declared elsewhere in our application.
+Here we are just declaring that such a connection will exist.
+
+```elixir
+  socket "/socket", HelloPhoenix.UserSocket
+```
+Next a series of plugs which would have to relevant to every route in your application.
+You may want to customize some of the features, enabling gzip:true when deploying to production
+with digested static files.
+
+Static files are served from priv/static before any part of our request makes it to a router.
+
+```elixir
+  plug Plug.Static,¬
+    at: "/", from: :hello_phoenix, gzip: false,
+    only: ~w(css fonts images js favicon.ico robots.txt)
+```
+A code reloading feature is included into our application which uses a socket
+to communicate that code can be reloaded for development preview.
```

Comment:
This isn't necessarily a development thing. Perhaps it could be rephrased as:

 > If code reloading is enabled, a socket will be used to communicate to the browser that the page needs to be reloaded when code is changed on the server. This feature is enabled by default in the development environment.